[SPARK-19799] Support WITH clause (CTE) in subqueries

gbloisi opened the pull request:
Because of SPARK-17590 support of WITH clause (CTE) in subqueries requires only grammar support.

Test for augmented syntax is provided.

gbloisi committed:
Because of SPARK-17590 support of WITH clause in subqueries requires only grammar support.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

gatorsmile commented:
ok to test

SparkQA commented:
**[Test build #98453 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98453/testReport)** for PR 22936 at commit [`66cd537`](https://github.com/apache/spark/commit/66cd5379a17e05707ae162bb20e9c64812737d78).

SparkQA commented:
**[Test build #98453 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98453/testReport)** for PR 22936 at commit [`66cd537`](https://github.com/apache/spark/commit/66cd5379a17e05707ae162bb20e9c64812737d78).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98453/
Test FAILed.

gbloisi committed:
Fix scala check issue 

FIX:
Scalastyle checks failed at following occurrences:
[error] /home/jenkins/workspace/SparkPullRequestBuilder/sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:545: File line length exceeds 100 characters

SparkQA commented:
**[Test build #98568 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98568/testReport)** for PR 22936 at commit [`5377087`](https://github.com/apache/spark/commit/5377087612e52eba09961e7591ae0c436c0a4de1).

SparkQA commented:
**[Test build #98568 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98568/testReport)** for PR 22936 at commit [`5377087`](https://github.com/apache/spark/commit/5377087612e52eba09961e7591ae0c436c0a4de1).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98568/
Test FAILed.

gbloisi committed:
Resolve scalastyle check error

FIX: [error] /home/jenkins/workspace/SparkPullRequestBuilder/sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:545:79: Whitespace at end of line

SparkQA commented:
**[Test build #98569 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98569/testReport)** for PR 22936 at commit [`3080ed4`](https://github.com/apache/spark/commit/3080ed4eab23b74bcafd7f933808b116402fe622).

SparkQA commented:
**[Test build #98569 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98569/testReport)** for PR 22936 at commit [`3080ed4`](https://github.com/apache/spark/commit/3080ed4eab23b74bcafd7f933808b116402fe622).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98569/
Test PASSed.

gatorsmile commented on sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:
```diff
@@ -535,7 +535,19 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
       Row(5, "5") :: Row(4, "4") :: Nil)
+  }
 
+  test("SPARK-19799: Support WITH clause in subqueries") {
```
Could you add more test cases? For example, can a with clause in a subquery shadow a with clause in an enclosing query with the same name? Another example, use with clauses in a subquery expression?

gbloisi replied:
I need to find some spare time for that, but I'll surely do sooner or later. I assumed that most of the correctness comes from https://github.com/apache/spark/pull/15146 but surely there may be interactions that goes beyond that.

maropu commented:
@gbloisi Any update?

gatorsmile commented:
@chrysanxia Could you take over this?

peter-toth commented:
@gatorsmile @gbloisi @chrysanxia may I take over this?
I played a bit with recursion support recently:
https://github.com/apache/spark/pull/23531# and during porting https://github.com/postgres/postgres/blob/master/src/test/regress/sql/with.sql I also found that WITH is not supported in subqueries https://github.com/apache/spark/pull/23531#discussion_r249968739 (point 6).

peter-toth commented:
@gatorsmile here is my PR: https://github.com/apache/spark/pull/24831
